src/paths.rs

```diff
@@ -0,0 +1,583 @@
+        } else if !graph[self.end_node].is_definition() {
+            return false;
+        } else if !self.symbol_stack.is_empty() {
+            return false;
+        } else if !self.scope_stack.is_empty() {
+            return false;
+        } else {
+            true
+        }
+    }
+
+    pub fn display<'a>(&'a self, graph: &'a StackGraph, paths: &'a mut Paths) -> impl Display + 'a {
+        display_with(self, graph, paths)
+    }
+}
+
+impl<'a> DisplayWithPaths for &'a Path {
+    fn prepare(&mut self, graph: &StackGraph, paths: &mut Paths) {
+        self.symbol_stack.clone().prepare(graph, paths);
+        self.scope_stack.clone().prepare(graph, paths);
+    }
+
+    fn display_with(
+        &self,
+        graph: &StackGraph,
+        paths: &Paths,
+        f: &mut std::fmt::Formatter,
+    ) -> std::fmt::Result {
+        write!(
+            f,
+            "{} -> {}",
+            self.start_node.display(graph),
+            self.end_node.display(graph),
+        )?;
+        if !self.symbol_stack.is_empty() || !self.scope_stack.is_empty() {
+            write!(
+                f,
+                " <{}> ({})",
+                display_prepared(self.symbol_stack, graph, paths),
+                display_prepared(self.scope_stack, graph, paths),
+            )?;
+        }
+        Ok(())
+    }
+}
+
+/// Errors that can occur during the path resolution process.
+#[derive(Debug)]
+pub enum PathResolutionError {
+    /// The path contains a _jump to scope_ node, but there are no scopes on the scope stack to
+    /// jump to.
+    EmptyScopeStack,
+    /// The path contains a _pop symbol_ or _pop scoped symbol_ node, but there are no symbols on
+    /// the symbol stack to pop off.
+    EmptySymbolStack,
+    /// The path contains a _pop symbol_ or _pop scoped symbol_ node, but the symbol at the top of
+    /// the symbol stack does not match.
+    IncorrectPoppedSymbol,
+    /// The path contains an edge whose source node does not match the sink node of the preceding
+    /// edge.
+    IncorrectSourceNode,
+    /// The path contains a _pop scoped symbol_ node, but the symbol at the top of the symbol stack
+    /// does not have an attached scope list to pop off.
+    MissingAttachedScopeList,
+    /// The path contains a _pop symbol_ node, but the symbol at the top of the symbol stack has an
+    /// attached scope list that we weren't expecting.
+    UnexpectedAttachedScopeList,
+}
+
+impl Path {
+    /// Attempts to append an edge to the end of a path.  If the edge is not a valid extension of
+    /// this path, we return an error describing why.
+    pub fn append(
+        &mut self,
+        graph: &StackGraph,
+        paths: &mut Paths,
+        edge: Edge,
+    ) -> Result<(), PathResolutionError> {
+        if edge.source != self.end_node {
+            return Err(PathResolutionError::IncorrectSourceNode);
+        }
+
+        let sink = &graph[edge.sink];
+        if let Node::PushSymbol(sink) = sink {
```

Comment:
Would a `match` statement be more readable here than a chain of if-lets?

Reply:
It aesthetics, but I tend to not like `match` statements when the bodies are large like this.